### ISSUE TITLE
Search results can still be improved, again

### ISSUE THREAD

**Issue by heathermiller** [scalacenter/scaladex#211]
Despite opening many tickets about this by this point, the search relevance still doesn't make much sense in many normal usage scenarios. 

For example, I just searched for a new project that hasn't yet been published on Sonatype. It's called "lchannels". I get 3162 results that have no relationship whatsoever to this library.

If there is no library published under this name, I shouldn't be shown more than 3000 random results. There's no way that "lchannels" appears in the README, dependencies, or any other field that could be indexed. So I don't understand what's being shown – it seems completely random.

**Comment by MasseGuillaume** [comment]
solved in 40f13ed60ba12c28bdab9da58bcebe7f07c3b5cc

let's chat before opening issues

**Comment by heathermiller** [comment]
Opening issues is a way of chatting :)

**Comment by heathermiller** [comment]
I would strongly suggest not to discourage feedback in the form of issues. It discourages people from reporting regressions and sharing suggestions for usage improvements.

**Comment by olafurpg** [comment]
I would argue that this is still an open issue. For example, searching ["parser"](https://index.scala-lang.org/search?q=parser) yields a lot of random projects and searching ["parse"](https://index.scala-lang.org/search?q=parse) yields an different set of projects (1 project in common with "parser"). I think that the search index should list similar projects for those two queries, from the user's perspective they're identical queries.

**Comment by MasseGuillaume** [comment]
we can search with a distance:
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-fuzzy-query.html

we can also provide "did you mean X?"
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-suggesters-phrase.html

**Comment by heathermiller** [comment]
Also, today @xeno-by pointed out that when searching for scala-reflect, he found 3182 entries and wasn't sure what to make of it. Surely that wasn't dependencies...

**Comment by xeno-by** [comment]
Here's the screenshot: https://www.dropbox.com/s/d28adc4yqwaoz9w/Screenshot%202016-08-02%2015.02.17.png?dl=0.

**Comment by MasseGuillaume** [comment]
so Scala reflect is part of the scala/scala project that's why it's harder to find

http://search.maven.org/#artifactdetails%7Corg.scala-lang%7Cscala-reflect%7C2.11.8%7Cjar

**Comment by heathermiller** [comment]
Just two thoughts:

1. Valid point, but still, should it list 3182 results? If an artifact called "scala-reflect" isn't found, shouldn't we return zero results?
2. It probably should be easy to find scala-reflect at least... Maybe even scala-library.

**Comment by MasseGuillaume** [comment]
yeah I have an easy idea to solve this ;-).

**Comment by olafurpg** [comment]
I still don't get the "most relevant" ordering. Here's the results for ["pars"](https://index.scala-lang.org/search?q=pars)

<img width="1182" alt="screen shot 2016-08-10 at 23 04 13" src="https://cloud.githubusercontent.com/assets/1408093/17570992/d38a9e04-5f4e-11e6-9d12-89bc530d9dd2.png">

In my opinion, the "ascii-graphs" library is most relevant out of the top 5 libraries (scala/scala-parser-combinators or fastparse should really be the top result though). I'm guessing gressy-entities is most relevant because it has the most recent commit. If that's the case, then I think that fact should be highlighted on the result page with a "last updated X ago" message. However, I still feel that stars/forks/contributors should weigh more than most recent commit.

**Comment by MasseGuillaume** [comment]
There is a fuzzy match on the description. Part -> Pars that's why it's on top. I think in this case we are not talking about search relevance but typo fixing. Google will show you something like: *Did you mean parse?* Elasticsearch has an api for this:
https://www.elastic.co/guide/en/elasticsearch/reference/2.3/search-suggesters.html I will take a look.

Another thing I can improve related to Elasticsearch is to turn on the analyzer for the artifacts, keywords, etc. That's why it's so strict now. This is [easy to fix](https://github.com/CodeBrew-io/Sauce/blob/develop/web/app/model/Snippet.scala#L82-L85).

A longer task I have in mind is to put more emphasis on the artifacts. Right now you search for a project. Searching for `reflect` would yield scala/scala and that makes it less obvious why it's relevant.

In short, relevance is a hard thing to figure out, I think we will improve over time.

**Comment by olafurpg** [comment]
> not talking about search relevance but typo fixing. Google will show you something like: Did you mean parse? 

Correct. "Did you mean ...?"  would be a really useful feature.

>  turn on the analyzer for the artifacts, keywords, etc

It would be interesting to see if that could improve the search results.

> In short, relevance is a hard thing to figure out, I think we will improve over time.

I can imagine it's hard, especially since users expect Google-like quality in the search results. I'm sorry for giving you such a hard time. The search has made a lot of progress since June, big kudos for that.
